agasc/scripts/update_mag_supplement.py

```diff
@@ -23,35 +27,61 @@ def parser():
                             ' CxoTime-compatible time stamp.'
                             ' Default: now.')
     parse.add_argument('--whole-history',
-                       help='Include all star observations.')
-    parse.add_argument('--obs-status-override', help='YAML file with observation status.')
-    parse.add_argument('--obs', help='Observation ID for status override.')
-    parse.add_argument('--agasc-id', help='AGASC ID for status override.')
-    parse.add_argument('--status', help='Status to override.')
-    parse.add_argument('--comments', help='Comments for status override.', default='')
-    parse.add_argument('--email', help='Email to report errors.')
-    parse.add_argument('--report',
-                       help='Generate HTML report for the period covered. Default: False',
+                       help='Include all star observations and ignore --start/stop.',
                        action='store_true', default=False)
+    parse.add_argument('--agasc-id-file',
+                       help='Include only observations of stars whose AGASC IDs are specified '
+                            'in this file, one per line.')
     parse.add_argument('--output-dir',
-                       help='Directory where to place the supplement. Default: .',
+                       help='Directory where agasc_supplement.h5 is located.'
+                            'Other output is placed here as well. Default: .',
                        default='.')
-    parse.add_argument('--reports-dir',
-                       help='Directory where to place reports.'
-                            ' Default: <output_dir>/supplement_reports/weekly.')
-    parse.add_argument('--multi-process',
+    parse.add_argument('--include-bad',
+                       help='Do not exclude "bad" stars from magnitude estimates. Default: False',
+                       action='store_true', default=False)
+    status = parse.add_argument_group(
```

Comment:
I think there is some misunderstanding here.

- the actual update of the OBS status is done in only one place, even if it is called from several scripts (nice thing, modularity...).
- Maybe there is a way to define the parser only once.

I think it is useful to have this here. When you override the status of an observation, you usually want to update the magnitude right away. If you do not have these options, then you have to keep track of which observation you updated, and make sure the magnitude is recalculated. The way it is, the recalculation happens automatically.

Same with reports: The reporting function is one function. Calling it from this script allows you to include the differences between magnitudes in the supplement before/after updating. This shows up in the report, so you can see whether updates to the supplement were large. You can't do this if you have them in different scripts.

Again: no code is repeated. I am just taking advantage of the modularity to add this convenient behavior to the update script.

Reply:
I just checked in a version where I removed the bad-stars script (merged into the update-obs-status one) and made it so the mag-stats-update script can take the same command-line arguments of the update-obs-status script.

Everything is contained within update-obs-status script, so the mag-stats update almost does not need to know any detail of it. I say "almost" because it can't add a command-line argument that is already added by the update-obs-status script, but that's it.

Reply:
We seem to have a difference of style here, as I prefer "one script does one thing" and you seem to prefer "one script does everything". That's OK, but we should not have a mix of the two styles. So if you want one script `agasc-update-supplement` that does everything, then just one script.

>  When you override the status of an observation, you usually want to update the magnitude right away.

This is not how I envision our production process. I think the ACA reviewer process should be:
- Look at the weekly report
- Make dispositions on any suspect stars
- Run a command or commands to update the `obs_status` or `bad` tables. This is fast.
- Do not run the mag estimate update nor make new reports. This takes a long time and more importantly it should be done only by the weekly cron job so that task logging and error checking gets done.

Related to scripts, the task_schedule.cfg file now has this, which makes me confused because it looks redundant wrt making the report.
```
      exec 1: agasc-supplement-update --output-dir /proj/sot/ska/data/agasc/rc --report
      exec 1: agasc-mag-estimate-report --output-dir /proj/sot/ska/data/agasc/rc
```

Reply:
> Do not run the mag estimate update nor make new reports. This takes a long time and more importantly it should be done only by the weekly cron job so that task logging and error checking gets done.

But the thing is: there is no guarantee that the weekly run will catch on the change and produce a new report. If you do not mind having stale reports, then there is no problem.

> the task_schedule.cfg file now has this, which makes me confused because it looks redundant wrt making the report.

That's because there are actually two reports being produced:
- the weekly run
- a running 60 day collection of suspect observations.

Reply:
also, in relation to your process. In your process, the weekly would not show the difference in magnitude between the previous and current versions of the supplement (to see whether something had a large change). If you do not want this difference shown, then there is no problem.

I would argue this is not a matter of "style" where I somehow want a script to do "everything" (you are the one who has an opinion on style). I want the two things I just mentioned, but if you do not like having them, then I can do it the way you want.

Reply:
> Calling it from this script allows you to include the differences between magnitudes in the supplement before/after updating. This shows up in the report ...

Fair enough. If the report generation is tightly coupled to the update process in that way, then I don't see the point in there being a separate `agasc-mag-estimate-report` script since it can't really do the job. Unless I'm not understanding something here.

Reply:
> But the thing is: there is no guarantee that the weekly run will catch on the change and produce a new report. If you do not mind having stale reports, then there is no problem.

Stale/inaccurate reports are not OK.  Maybe we need a different process, where the reviewer puts a YAML file with the dispositions into a predefined location and the weekly cron picks that up and processes accordingly. 

Is that guaranteed to always work, even for a star that has not been observed recently (or within the standard 2-week processing window)?

Reply:
I feel like we are going in circles.

> If the report generation is tightly coupled to the update process in that way, then I don't see the point in there being a separate agasc-mag-estimate-report

there is a point in having a separate agasc-mag-estimate-report, because we actually produce two reports. One with the diffs (the weekly report) and one without (the suspect obs).

> Stale/inaccurate reports are not OK. Maybe we need a different process, where the reviewer puts a YAML file with the dispositions

Placing a yaml file is not a problem. In the process you suggest, one first updates the supplement (say, from a yaml file), and then one has to give the same yaml file to the update+report to make sure it is updated. If you are anyway giving it as an argument to the script, you might as well do the update. All the functionality is modular and decoupled. It is literally one line of code: https://github.com/sot/agasc/blob/update-script-fixes/agasc/scripts/update_mag_supplement.py#L98

Reply:
Sorry, I think part of my issue is not having understood the whole process well enough. I'm getting there, but I originally believed that each part could be run as a separate script, so it seemed like a design choice to make all the functionality accessible from one script. Now I understand that was not correct.

This makes sense given the original concept that there is basically just the one set of files, so the code cannot know what changed. Since we have somewhat formally adopted the idea of a flight and rc version of the supplement, I wonder if those two versions might be sufficient to allow decoupling updates of obs_status/bad, mag estimates, and report generation. I.e. if the code can know what changed from the files instead of from variables in memory.

But anyway, it's not worth changing right now if we can come up with a workable process.

Reply:
That's right, and it relates to Jean's comment about being able to reproduce star selection with a given version of the supplement. I never thought of setting it up as a table that can change in time, and that shows.

In principle, we could add a historical table to hold all values. The way it works is that you never remove or edit records from the history, you only add, and the historical table sits side-by-side with the main table. That allows you to reconstruct the state at any time. We could also add a "last_modified" timestamp, which does not allow you to reconstruct the table at any time, but at least you know when something changed.

Reply:
With regard to reproducing star selection, I think Tom's idea to use a local github repo as the "promotion storage" thing is fine. So we'd have somewhat "private" versioned files if we needed to look something up.

If we need a mechanism to more efficiently re-calculate the mag stats as they would have existed at a certain moment in time, I think the code and storage could be re-organized to accommodate that ... but I don't really see a driver.

Reply:
The more I think the more I see the original design concept and the need for the YAML file for the primary way of introducing disposition updates. So I do believe that the existing code is basically workable, with this small change:
- The cron task for the weekly mag estimate update and report generation should be always looking for an obs-status update YAML file.
- That file could be something like `$SKA/data/agasc/rc/star-status-pending.yml` (or whatever).
- It would be optional.
- The cron task schedule would then remove that file upon successful completion of the update processing, i.e. `rm -f $ENV{SKA}/data/agasc/rc/star-status-pending.yml` as an exec after the update script.

Thoughts?

Reply:
I didn't explicitly mention the crux point that the ACA reviewer process includes putting that YAML file into the correct place each week.